### PR TITLE
split user time on task form and remove timeentry models

### PR DESCRIPTION
**Pull request by stuartaccent**
split time by user

image

remove the tasktiming models in favour of view so craigs bits are ok

### CODE REVIEW

File: frontend/src/app/state/selectors/task.ts [hunk 1]
```diff
@@ -37,7 +27,12 @@ export const getTaskCollection = createSelector(
             return _.assign({}, task, {
                 _job: _.find(jobs, ['id', task.job]),
                 _status: _.find(statuses, ['id', task.status]),
-                _user: _.find(users, ['id', task.created_by])
+                _user: _.find(users, ['id', task.created_by]),
+                _timing_summary: {
+                    allocated_hours: _.sumBy(task.timings, (item: any) => +item.allocated_hours).toFixed(2),
+                    time_spent_hours: _.sumBy(task.timings, (item: any) => +item.time_spent_hours).toFixed(2),
+                    is_over_allocated_hours: _.sumBy(task.timings, (item: any) => +item.time_spent_hours) > _.sumBy(task.timings, (item: any) => +item.allocated_hours)
+                }
```
Comment (craigvincent) [inline frontend/src/app/state/selectors/task.ts:35]: I assume these are cached in the state container so pretty efficient during app use - but does the `sumBy` work cause any delay when initialising the app?

File: wip/migrations/0035_auto_20220120_1533.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+            ta.job_id,
+            SUM(COALESCE(tm.allocated_hours, 0)) AS allocated_hours,
+            SUM(COALESCE(tm.time_spent_hours, 0)) AS time_spent_hours
+            FROM wip_task ta
+            left join public.view_tasktiming tm 
+            on ta.id = tm.task_id 
+            GROUP by
+            ta.job_id;
+            """
+        ),
```
Comment (craigvincent) [inline wip/migrations/0035_auto_20220120_1533.py:47]: Said it to you already, but do like the idea of these being views - and nice they are tracked in the migrations :)

File: frontend/src/app/state/effects/task.ts [hunk 3]
```diff
@@ -13,9 +18,18 @@ export class TaskEffects extends APIBaseEffects {
...
+        map((action: IActionWithPayload) => action.payload),
+        mergeMap(id =>
+            this.service$.one(this.apiUrl, id).pipe(
+                switchMap(data => [
+                    ({type: `${this.prefix} LOAD_ONE_SUCCESS`, payload: data}),
+                    ({type: `[Job] LOAD_ONE`, payload: data.job}),
+                ]),
+                catchError(res => of({type: HttpActions.HTTP_ERROR, payload: res}))
+            )
+        )
```
Comment (craigvincent) [inline frontend/src/app/state/effects/task.ts:32]: Not going to pretend I fully understand this in the slightest 😉 `mergeMap` allows multiple requests to continue concurrently?

**Review by craigvincent — APPROVED** [review]
> so craigs bits are ok
😆 

It almost feels slightly tidier! From looking through it looks sensible. 

Anything else to worry about with regards to moneyworks exports etc that rely on time?

File: frontend/src/app/state/effects/task.ts (hunk 3 above)
Comment (stuartaccent) [reply frontend/src/app/state/effects/task.ts:32]: good spot, it is meant to be switchMap

File: frontend/src/app/state/selectors/task.ts (hunk 1 above)
Comment (stuartaccent) [reply frontend/src/app/state/selectors/task.ts:35]: i would doubt it they are in selectors that will only run when the input data changes, the timings are already grouped to the user on each task so there is only 1 row per user who is either assigned or spent any time on it

File: wip/migrations/0035_auto_20220120_1533.py (hunk 2 above)
Comment (stuartaccent) [reply wip/migrations/0035_auto_20220120_1533.py:47]: yeh its not sommit ive used too much in theses things but do like it esp when you can still define a table in the code for it

File: frontend/src/app/state/selectors/task.ts (hunk 1 above)
Comment (stuartaccent) [reply frontend/src/app/state/selectors/task.ts:35]: i cant find any other ref to these models so should be ok, they were only ever really there so the front end didnt need to get all the tasks in the first place, was amazed to see it in the reports tbh

**Pull request merged**